File: samples/bot-daily-task-reminder/nodejs/bots/teamsBot.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

const { TeamsActivityHandler, CardFactory, TurnContext } = require("botbuilder");
const { TaskModuleResponseFactory } = require("../models/taskModuleResponseFactory");
const schedule = require('node-schedule');
const conversationReferences = {};
let adapter;

/**
 * TeamsBot class handles Teams activities and task modules.
 */
class TeamsBot extends TeamsActivityHandler {
    constructor() {
        super();
        this.baseUrl = process.env.BaseUrl;

        // Handle when a new member is added to the conversation.
        this.onMembersAdded(async (context, next) => {
            const membersAdded = context.activity.membersAdded;
            for (const member of membersAdded) {
                if (member.id !== context.activity.recipient.id) {
                    await context.sendActivity("Hello and welcome! With this sample, you can schedule a recurring task and receive reminders at the scheduled time. Use the command 'create-reminder' to start.");
                }
            }
            await next();
        });

        // Handle incoming messages.
        this.onMessage(async (context, next) => {
            if (context.activity.text.toLowerCase().trim() === "create-reminder") {
                const userCard = CardFactory.adaptiveCard(this.adaptiveCardForTaskModule());
                await context.sendActivity({ attachments: [userCard] });
            }
            await next();
        });
    }

    /**
     * Handle task module fetch.
     * @param {TurnContext} context - The context object for the turn.
     * @param {Object} taskModuleRequest - The task module request object.
     * @returns {Object} - The task module response.
     */
    handleTeamsTaskModuleFetch(context, taskModuleRequest) {
        const cardTaskFetchId = taskModuleRequest.data.id;

        if (cardTaskFetchId === "schedule") {
            return TaskModuleResponseFactory.toTaskModuleResponse({
                url: `${this.baseUrl}/scheduleTask`,
                fallbackUrl: `${this.baseUrl}/scheduleTask`,
                height: 350,
                width: 350,
                title: "Schedule a task"
            });
        }
        return null;
    }

    /**
     * Handle task module submit action.
     * @param {TurnContext} context - The context object for the turn.
     * @param {Object} taskModuleRequest - The task module request object.
     */
    async handleTeamsTaskModuleSubmit(context, taskModuleRequest) {
        const taskDetails = {
            title: taskModuleRequest.data.title,
            dateTime: taskModuleRequest.data.dateTime,
            description: taskModuleRequest.data.description,
            selectedDays: taskModuleRequest.data.selectedDays,
        };

        this.saveTaskDetails(taskDetails);
        await context.sendActivity("Task submitted successfully, you will get a recurring reminder for the task at a scheduled time");

        const currentUser = context.activity.from.id;
        conversationReferences[currentUser] = TurnContext.getConversationReference(context.activity);
        adapter = context.adapter;

        const date = new Date(taskModuleRequest.data.dateTime);
        const cronExpression = `${date.getMinutes()} ${date.getHours()} * * ${taskModuleRequest.data.selectedDays.toString()}`;

        schedule.scheduleJob(cronExpression, async () => {
            try {
                const botAppId = process.env.MicrosoftAppId || process.env.AAD_APP_CLIENT_ID || '';
                if (!botAppId) {
                    console.warn('MicrosoftAppId is not set in environment. Proactive send may fail.');
                }
                await adapter.continueConversationAsync(botAppId, conversationReferences[currentUser], async turnContext => {
                    const userCard = CardFactory.adaptiveCard({
                    $schema: "http://adaptivecards.io/schemas/adaptive-card.json",
                        body: [
                            {
                            type: "TextBlock",
                            size: "Default",
                            weight: "Bolder",
                            text: "Reminder for a scheduled task!"
                            },
                            {
                            type: "TextBlock",
                            size: "Default",
                            weight: "Default",
                                text: `Task title: ${taskDetails.title}`,
                                wrap: true
                            },
                            {
                            type: "TextBlock",
                            size: "Default",
                            weight: "Default",
                                text: `Task description: ${taskDetails.description}`,
                                wrap: true
                            }
                        ],
                    type: "AdaptiveCard",
                    version: "1.2"
                    });

                    await turnContext.sendActivity({ attachments: [userCard] });
                });
            }
            catch (err) {
                console.error('Error sending proactive reminder:', err && err.stack ? err.stack : err);
            }
        });

        return null;
    }

    /**
     * Save task details.
     * @param {Object} taskDetails - The task details object.
     */
    saveTaskDetails(taskDetails) {
        this.taskDetails = taskDetails;
    }

    /**
     * Create the adaptive card for the task module.
     * @returns {Object} - The adaptive card object.
     */
    adaptiveCardForTaskModule = () => ({
        $schema: "http://adaptivecards.io/schemas/adaptive-card.json",
        body: [
            {
                type: "TextBlock",
                size: "Default",
                weight: "Bolder",
                text: "Please click on schedule to schedule a task"
            },
            {
                type: "ActionSet",
                actions: [
                    {
                        type: "Action.Submit",
                        title: "Schedule task",
                        data: {
                            msteams: {
                                type: "task/fetch"
                            },
                            id: "schedule"
                        }
                    }
                ]
            }
        ],
        type: "AdaptiveCard",
        version: "1.2"
    });
}

module.exports.TeamsBot = TeamsBot;

File: samples/TeamsSDK/Archived/bot-people-picker-adaptive-card/python/bots/teamsBot.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

from botbuilder.core.teams import TeamsActivityHandler
from botbuilder.schema import Activity
from botbuilder.schema.teams import TeamInfo, TeamsChannelAccount
from botbuilder.core import TurnContext
from adaptive_cards import cards

class TeamsBot(TeamsActivityHandler):
    def __init__(self):
        # Constructor for the TeamsBot class. Initializes the parent class (TeamsActivityHandler)
        super().__init__()

    async def on_teams_members_added(
        self,
        teams_members_added: [TeamsChannelAccount],
        team_info: TeamInfo,
        turn_context: TurnContext,
    ):
        """
        Handles the event when new members are added to a team.
        
        Args:
            teams_members_added: List of members added to the team.
            team_info: Information about the team where the members were added.
            turn_context: Context object for the current turn of the conversation.
        
        Behavior:
            Sends a welcome message to newly added members, excluding the bot itself.
        """
        for member in teams_members_added:
            # Check if the added member is not the bot itself
            if member.id != turn_context.activity.recipient.id:
                await turn_context.send_activity(
                    f"Hello and welcome! With this sample you can see the functionality of people-picker in adaptive card"
                )

    async def on_message_activity(self, turn_context: TurnContext):
        TurnContext.remove_recipient_mention(turn_context.activity)

        if turn_context.activity.text:
            if turn_context.activity.conversation.conversation_type == "personal":
                user_card = cards.adaptive_card_for_personal_scope()
                await turn_context.send_activity(Activity(attachments=[user_card]))
            else:
                user_card = cards.adaptive_card_for_channel_scope()
                await turn_context.send_activity(Activity(attachments=[user_card]))
        elif turn_context.activity.value:
            task_details = turn_context.activity.value
            await turn_context.send_activity(
                f"Task title: {task_details['taskTitle']},\n"
                f"Task description: {task_details['taskDescription']},\n"
                f"Task assigned to: {task_details['userId']}"
            )

